src/WebRTC/SessionDescriptionHandler.js

```diff
@@ -0,0 +1,446 @@
+   * @param {String} contentType The content type that is in the SIP Message
+   * @returns {boolean}
+   */
+  hasDescription: {writable: true, value: function hasDescription (contentType) {
+    return contentType === this.CONTENT_TYPE;
+  }},
+
+  /**
+   * The modifier that should be used when the session would like to place the call on hold
+   * @param {String} [sdp] The description that will be modified
+   * @returns {Promise} Promise that resolves with modified SDP
+   */
+  holdModifier: {writable: true, value: function holdModifier (sdp) {
+    if (!(/a=(sendrecv|sendonly|recvonly|inactive)/).test(sdp)) {
+      sdp = sdp.replace(/(m=[^\r]*\r\n)/g, '$1a=sendonly\r\n');
+    } else {
+      sdp = sdp.replace(/a=sendrecv\r\n/g, 'a=sendonly\r\n');
+      sdp = sdp.replace(/a=recvonly\r\n/g, 'a=inactive\r\n');
+    }
+    return SIP.Utils.Promise.resolve(sdp);
+  }},
+
+  /**
+   * Set the remote description to the underlying media implementation
+   * @param {String} sessionDescription The description provided by a SIP message to be set on the media implementation
+   * @param {Object} [options] Options object to be used by getDescription
+   * @param {MediaStreamConstraints} [options.constraints] MediaStreamConstraints https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamConstraints
+   * @param {Object} [options.peerConnectionOptions] If this is set it will recreate the peer connection with the new options
+   * @param {Array} [modifiers] Array with one time use description modifiers
+   * @returns {Promise} Promise that resolves once the description is set
+   */
+  setDescription: {writable:true, value: function setDescription (sessionDescription, options, modifiers) {
+    var self = this;
+
+    options = options || {};
+    if (options.peerConnectionOptions) {
+      this.initPeerConnection(options.peerConnectionOptions);
+    }
+
+    // Merge passed constraints with saved constraints and save
+    this.constraints = Object.assign(this.constraints, options.constraints);
+    this.constraints = this.checkAndDefaultConstraints(this.constraints);
+
+    modifiers = modifiers || [];
+    if (!Array.isArray(modifiers)) {
+      modifiers = [modifiers];
+    }
+    modifiers = modifiers.concat(this.modifiers);
+
+    return SIP.Utils.reducePromises(modifiers, sessionDescription)
+      .then(function(sdp) {
+        var rawDescription = {
+          type: self.hasOffer('local') ? 'answer' : 'offer',
+          sdp: sdp
+        };
+        // TODO: This emit does not match what we are doing for getDescription... kind of re: modifier
+        self.emit('setDescription', rawDescription);
+
+        return self.peerConnection.setRemoteDescription(new self.WebRTC.RTCSessionDescription(rawDescription));
+      })
+      .catch(function modifierError(e) {
+        self.logger.error("The modifiers did not resolve successfully");
+        self.logger.error(e);
+        throw e;
```

Comment:
We should log the SDP that was failed to set. It would be helpful for debugging. Also if it was an offer or answer.